### PR TITLE
[QoI] Improve diagnostics for value of meta types

### PR DESCRIPTION
**Pull request by rintaro**
* Suggest adding `()` only if it has accessible constructors.
* When adding `.self` to non-`isSimple()` type expressions, enclose it with parenthesis;
  e.g `(Int) -> Self` or `P1 & P2`

### CODE REVIEW

File: lib/Sema/MiscDiagnostics.cpp [hunk 1]
```diff
@@ -539,18 +545,25 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
 
       TC.diagnose(E->getStartLoc(), diag::value_of_metatype_type);
 
-      // Add fix-t to insert '()', unless this is a protocol metatype.
-      bool isProtocolMetatype = false;
+      // Add fix-t to insert '()', only if this is a metatype of
```
Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: Drive-by fix: "fix-t".

File: lib/Sema/MiscDiagnostics.cpp [hunk 2]
```diff
@@ -539,18 +545,25 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
...
       // Add fix-it to insert ".self".
-      TC.diagnose(E->getEndLoc(), diag::add_self_to_type)
-        .fixItInsertAfter(E->getEndLoc(), ".self");
+      auto diag = TC.diagnose(E->getEndLoc(), diag::add_self_to_type);
+      if (E->canAppendCallParentheses())
+        diag.fixItInsertAfter(E->getEndLoc(), ".self");
+      else
+        diag
+          .fixItInsert(E->getStartLoc(), "(")
+          .fixItInsertAfter(E->getEndLoc(), ").self");
```
Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: Style nitpick: we usually don't bother chaining once the diagnostic has a name.

```
diag.fixItInsert(E->getStartLoc(), "(");
diag.fixItInsertAfter(E->getEndLoc(), ").self");
```

(and we'd probably use braces around the multi-line `else` either way, even though it's technically one statement this way, and then do the `if` branch as well for symmetry.)

**Review by jrose-apple — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test and merge
